### PR TITLE
Implement code mappings

### LINKED ISSUE #1647: Create proper code mappings for procedural macro results
After expanding Cairo code with procedural macros, construct `code_mappings` for new `PluginGeneratedFile`, between original code span saved in each token of expanded `TokenStream` and new position in outputted file.

### PR DESCRIPTION
**Pull request by DelevoXDG**
Closes #1647

### CODE REVIEW

**Review by wawel37 — CHANGES_REQUESTED** [review]

File: scarb/src/compiler/plugin/proc_macro/host/derive.rs [hunk 1]
```diff
@@ -99,6 +105,14 @@ impl ProcMacroHostPlugin {
                 continue;
             }
 
+            let mut mappings = generate_code_mappings(&result.token_stream);
+            for mapping in &mut mappings {
+                mapping.span.start = mapping.span.start.add_width(current_width);
+                mapping.span.end = mapping.span.end.add_width(current_width);
+            }
+            code_mappings.extend(mappings);
+            current_width = current_width + TextWidth::from_str(&result.token_stream.to_string());
```
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/src/compiler/plugin/proc_macro/host/derive.rs:114]: Yes, as all derive macro expansions go into one file, so we need to adjust subsequent mappings.

File: scarb/src/compiler/plugin/proc_macro/host/derive.rs [hunk 2]
```diff
@@ -99,6 +105,14 @@ impl ProcMacroHostPlugin {
                 continue;
             }
 
+            let mut mappings = generate_code_mappings(&result.token_stream);
+            for mapping in &mut mappings {
+                mapping.span.start = mapping.span.start.add_width(current_width);
+                mapping.span.end = mapping.span.end.add_width(current_width);
```
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/src/compiler/plugin/proc_macro/host/derive.rs:111]: I think there's a significant difference between these two 😄 
```
mapping.span.start = mapping.span.start.add_width(current_width);
mapping.span.end = mapping.span.end.add_width(current_width);
``` 
```
mappings.span = generate_code_mappings_with_offset
```

I'm not sure it's worth it to add `generate_code_mappings_with_offset` that won't be used anywhere else

File: scarb/tests/build_cairo_plugin.rs [hunk 3]
```diff
@@ -1536,3 +1536,196 @@ fn can_expand_impl_inner_func_attrr() {
...
+        .current_dir(&project)
+        .assert()
+        .failure()
+        .stdout_matches(indoc! {r#"
+            [..] Compiling some v1.0.0 ([..]Scarb.toml)
+            [..] Compiling hello v1.0.0 ([..]Scarb.toml)
+            error: Cannot assign to an immutable variable.
+             --> [..]lib.cairo:4:1
+                x = 2;
+            ^********^
```
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/tests/build_cairo_plugin.rs:1584]: > @maciektr, is https://github.com/software-mansion/scarb/pull/1761 fixing that?

Yes, will rebase

File: scarb/src/compiler/plugin/proc_macro/host/mod.rs [hunk 4]
```diff
@@ -259,3 +263,33 @@ impl ProcMacroHost {
...
+                mappings.push(CodeMapping {
+                    span: TextSpan {
+                        start: current_pos,
+                        end: current_pos.add_width(token_width),
+                    },
+                    origin: CodeOrigin::Span(TextSpan {
+                        start: TextOffset::default()
+                            .add_width(TextWidth::new_for_testing(token.span.start as u32)),
+                        end: TextOffset::default()
+                            .add_width(TextWidth::new_for_testing(token.span.end as u32)),
```
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/src/compiler/plugin/proc_macro/host/mod.rs:285]: > You can check out @maciektr 's changes to the `TextWidth` API [here](https://github.com/starkware-libs/cairo/pull/6717#pullrequestreview-2458563771). Maybe you can use it as soon as it's merged

That's the plan!

File: scarb/tests/build_cairo_plugin.rs [hunk 5]
```diff
@@ -1536,3 +1536,196 @@ fn can_expand_impl_inner_func_attrr() {
...
+                    }}
+                "#};
+
+                let token_stream = TokenStream::new(vec![TokenTree::Ident(Token::new(
+                  code.clone(),
+                    TextSpan {
+                        start: 0,
+                        end: code.len(),
+                    },
+                ))]);
```
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/tests/build_cairo_plugin.rs:1679]: The only valid test is the one related to attributes, the other two are just placeholder tests. In the future, we want to have something like:
```
error: The value does not fit within the range of type core::integer::u8.
 --> /private/var/folders/z3/j6qq2t4n0tn0ycjgt4ssnz740000gn/T/.tmpYBzxRQ/hello/src/lib.cairo:6:8
#[derive(CustomDerive, Drop)]
               ^************^  
  = note: this error originates in the derive macro `custom_derive`
```

For that to happen, #1749 and https://github.com/starkware-libs/cairo/pull/6729 need to be merged. We are planning to address this afterwards and decided to leave the tests as is, so we can keep track of what needs to be changed.

File: scarb/src/compiler/plugin/proc_macro/host/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/src/compiler/plugin/proc_macro/host/mod.rs]: Updated

File: Cargo.toml [hunk 6]
```diff
@@ -36,25 +36,26 @@ assert_fs = "1"
...
-cairo-lang-semantic = "2.9.1"
-cairo-lang-sierra = "2.9.1"
-cairo-lang-sierra-to-casm = "2.9.1"
-cairo-lang-starknet = "2.9.1"
-cairo-lang-starknet-classes = "2.9.1"
-cairo-lang-syntax = "2.9.1"
-cairo-lang-test-plugin = "2.9.1"
-cairo-lang-test-runner = "2.9.1"
-cairo-lang-utils = { version = "2.9.1", features = ["env_logger"] }
+cairo-lang-casm = { git = "https://github.com/starkware-libs/cairo", rev = "f30960c9393148e6fee7cb1333635fad3995a6a9" }
```
Comment (maciektr) [inline Cargo.toml:39]: Please use current Cairo main - even if it won't pass some tests (pls ignore/change asserts)

File: scarb/src/compiler/plugin/proc_macro/host/derive.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DelevoXDG) [reply scarb/src/compiler/plugin/proc_macro/host/derive.rs:114]: > I guess you can replace the for below with just this token_stream.to_string() call, right? 🤔

Ok

> I guess you can replace the for below with just this token_stream.to_string() call, right? 🤔 Also, if we only use `add_str` we can replace PatchBuilder with just a string 🤔

I suppose we can, but isn't there any specific reason you introduced it in the first place? 🤔

[the remark this answers is not in the document]
Comment (maciektr) [reply scarb/src/compiler/plugin/proc_macro/host/derive.rs:114]: > isn't there any specific reason you introduced it in the first place? 🤔

No idea 🤷‍♂️ Can't think of any good reason to do that.

File: Cargo.toml (hunk 6 above)
Comment (DelevoXDG) [reply Cargo.toml]: Ok, I reverted it, but that does kind of partially nerfs our ability to use mappings to generate correct diagnostics.

> Please use current Cairo main

Can you also clarify whether you meant main branch or the latest tag? 😅

Comment (maciektr) [reply Cargo.toml]: main branch :D 

> Ok, I reverted it, but without https://github.com/starkware-libs/cairo/pull/6717 we lose our ability to use mappings to generate correct diagnostics:

Yes, that's true :( 
We will update tests when we merge the needed changes on Cairo side

File: Cargo.toml [hunk 7]
```diff
@@ -36,6 +36,7 @@ assert_fs = "1"
 async-trait = "0.1"
 axum = { version = "0.6", features = ["http2"] }
 bumpalo = "3"
+cairo-lang-casm = "2.9.1"
```
Comment (maciektr) [inline Cargo.toml:39]: You don't seem to use it 😓

**Review by maciektr — APPROVED** [review]

**Review by wawel37 — APPROVED** [review]

**Pull request merged**